plugins/woocommerce/src/Blocks/Domain/Services/CheckoutFieldsSchema/Validation.php

```diff
@@ -0,0 +1,158 @@
+<?php
+declare( strict_types = 1);
+
+namespace Automattic\WooCommerce\Blocks\Domain\Services\CheckoutFieldsSchema;
+
+use Automattic\WooCommerce\Blocks\Domain\Services\CheckoutFieldsSchema\DocumentObject;
+use Opis\JsonSchema\{
+	Helper,
+	Validator
+};
+use Opis\JsonSchema\Errors\ErrorFormatter;
+use WP_Error;
+
+/**
+ * Service class validating checkout field schema.
+ */
+class Validation {
+	/**
+	 * Meta schema.
+	 *
+	 * @var string
+	 */
+	private static $meta_schema_json = '';
+
+	/**
+	 * Get the field schema with context.
+	 *
+	 * @param string $field_id The field ID.
+	 * @param array  $field_schema The field schema.
+	 * @param string $context The context.
+	 * @return array
+	 */
+	public static function get_field_schema_with_context( $field_id, $field_schema, $context ) {
+		$primary_key   = 'checkout';
+		$secondary_key = 'additional_fields';
+		switch ( $context ) {
+			case 'billing_address':
+			case 'shipping_address':
+				$primary_key   = 'customer';
+				$secondary_key = $context;
+				break;
+			case 'contact':
+				$primary_key   = 'customer';
+				$secondary_key = 'additional_fields';
+				break;
+		}
+		return [
+			$primary_key => [
+				'properties' => [
+					$secondary_key => [
+						'properties' => [
+							$field_id => $field_schema,
+						],
+					],
+				],
+			],
+		];
+	}
+
+	/**
+	 * Validate the field rules.
+	 *
+	 * @param DocumentObject $document_object The document object to validate.
+	 * @param array          $rules The rules to validate against.
+	 * @return bool|WP_Error
+	 */
+	public static function validate_document_object( DocumentObject $document_object, $rules ) {
+		$validator = new Validator();
+		$result    = $validator->validate(
+			Helper::toJSON( $document_object->get_data() ),
+			Helper::toJSON(
+				[
+					'$schema'    => 'http://json-schema.org/draft-07/schema#',
+					'type'       => 'object',
+					'properties' => $rules,
+				]
+			)
+		);
+
+		if ( ! $result->hasError() ) {
+			return true;
+		}
+
+		// Generate useful error message.
+		$formatter = new ErrorFormatter();
+		$formatted = $formatter->formatFlat( $result->error() );
+		$errors    = implode( ', ', count( $formatted ) > 2 ? array_slice( $formatted, 2 ) : $formatted );
```

Comment:
As discussed let's not return this, have a follow to parse over errors in exchange for hardcoded, translatable errors.

As a first pass we should either return the passed error or a generic error.

Reply:
https://github.com/woocommerce/woocommerce/pull/55047/commits/b28531a4a696064638d233927838205cbf53ff7f